src/rust/engine/fs/brfs/src/main.rs

```diff
@@ -742,25 +761,66 @@ async fn main() {
   }
   .expect("Error making store");
 
-  let _fs = mount(mount_path, store, runtime).expect("Error mounting");
-  loop {
-    std::thread::sleep(std::time::Duration::from_secs(1));
+  #[derive(Clone, Copy, Debug)]
+  enum Sig {
+    INT,
+    TERM,
+    Unmount,
   }
-}
 
-#[cfg(target_os = "macos")]
-fn unmount(mount_path: &str) -> i32 {
-  unsafe {
-    let path = CString::new(mount_path).unwrap();
-    libc::unmount(path.as_ptr(), 0)
+  fn install_handler<F>(install_fn: F, sig: Sig) -> impl StreamExt<Item = Option<Sig>>
+  where
+    F: Fn() -> SignalKind,
+  {
+    signal(install_fn())
+      .unwrap_or_else(|_| panic!("Failed to install SIG{:?} handler", sig))
+      .map(move |_| Some(sig))
   }
-}
 
-#[cfg(target_os = "linux")]
-fn unmount(mount_path: &str) -> i32 {
-  unsafe {
-    let path = CString::new(mount_path).unwrap();
-    libc::umount(path.as_ptr())
+  let sigint = install_handler(SignalKind::interrupt, Sig::INT);
+  let sigterm = install_handler(SignalKind::terminate, Sig::TERM);
+
+  match mount(mount_path, store, runtime.clone()) {
+    Err(err) => {
+      error!(
+        "Store {} failed to mount at {}: {}",
+        store_path, mount_path, err
+      );
+      std::process::exit(1);
+    }
+    Ok((_, receiver)) => {
+      match receiver.recv().unwrap() {
+        BRFSEvent::INIT => debug!("Store {} mounted at {}", store_path, mount_path),
+        BRFSEvent::DESTROY => {
+          warn!("Externally unmounted before we could mount.");
+          return;
+        }
+      }
+
+      let unmount = task::spawn_blocking(move || {
+        // N.B.: In practice recv always errs and we exercise the or branch. It seems the sender
+        // side thread always exits (which drops our BuildResultFS) before we get a chance to
+        // complete the read.
+        match receiver.recv().unwrap_or(BRFSEvent::DESTROY) {
+          BRFSEvent::DESTROY => Some(Sig::Unmount),
+          event => {
+            warn!("Received unexpected event {:?}", event);
+            None
+          }
+        }
+      })
+      .map(|res| res.unwrap())
+      .into_stream();
+
+      let mut shutdown_signal = sigint.merge(sigterm).merge(unmount).filter_map(|x| x);
```

Comment:
Nice. One alternative is to use https://docs.rs/futures/0.3.4/futures/future/fn.select_all.html to select on multiple futures without converting them into streams, but for the "I definitely only want the first result" case like this, this looks cleaner.

Reply:
The Signals only implement Stream so this would have to be the stream::select_all equivalent. I had tried that but ran into hard times getting a `vec![...]` of those three things that had a uniform type to satisfy `select_all`.